### PR TITLE
locate-dominating-file: init 0.0.1

### PR DESCRIPTION
**Pull request by roman**
###### Description of changes

This simple bash script utility allows bash developers to look up a specific file in the directory hierarchy. This tool becomes handy when searching for specific files like `.envrc` or `.git,` etc.

The name is inspired by the [`locate-dominating-file` elisp function](https://www.gnu.org/software/emacs/manual/html_node/elisp/Contents-of-Directories.html#index-locate_002ddominating_002dfile).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AndersonTorres** [comment]
This is a general-purpose tool, not restricted to Nix ecosystem. Indeed you cited an Emacs Lisp code that does the same.

I suggest you to create a Github repository for it before package it, instead of creating such an internalized package.

**Comment by roman** [comment]
> I suggest you to create a Github repository for it before package it, instead of creating such an internalized package.

Great idea. I created a [personal repository](https://github.com/roman/locate-dominating-file) for this project. 

I replicated the `default.nix` that repository has in `nixpkgs`, is the same content (different `src` attribute, though). 

Do you know if there is an elegant way to use that file instead? I'm guessing not, as the only way I see to make it work is that we do an `import from derivation`, and that probably is a _nono_.

**Comment by AndersonTorres** [comment]
> Do you know if there is an elegant way to use that file instead? I'm guessing not, as the only way I see to make it work is that we do an `import from derivation`, and that probably is a _nono_.

No. The "elegant way" to reuse them it is flake+overlay.

(You can look at https://github.com/nix-community/nixpkgs-wayland and https://github.com/nix-community/emacs-overlay as examples).

In the Nixpkgs monorepo we just ignore these extra Nix files that came from the upstream project. For us, they have more or less the same significance of a, say, a Makefile or a gitignore.

Certainly a well-written Nix file can be "copy-pasted", but this is not a rule.

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      repo   = "locate-dominating-file";
+      rev    = version;
+      sha256 = "sha256-gwh6fAw7BV7VFIkQN02QIhK47uxpYheMk64UeLyp2IY=";
+    };
+
+    doCheck = true;
+
+    postPatch = ''
+      for file in $(find src tests -type f); do
+        substituteInPlace "$file" --replace "/bin/bash" "${bash}/bin/bash"
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:26]: We use `patchShebangs <file>` to this.

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        substituteInPlace "$file" --replace "#!/usr/bin/env bash" "#!${bash}/bin/bash"
+      done
+    '';
+
+    buildInputs = [getopt];
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:31]:
```suggestion
    buildInputs = [ getopt ];
```

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 3]
```diff
@@ -0,0 +1,48 @@
+{
+  bats,
+  bash,
+  lib,
+  stdenv,
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:5]: Given that your package is a Bash shell script, you can use `stdenvNoCC` - a stdenv without C compiler.

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  getopt,
+  fetchFromGitHub,
+}: let
+  batsPkg = bats.withLibraries (p: [p.bats-support p.bats-assert]);
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:9]:
```suggestion
```

You can put this in checkInputs (see below).

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  version = "v0.0.1";
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:10]: `version` should start with a digit - except maybe `unstable-YYYY-MM-DD` (when picking a particular VCS commit).

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+in
+  stdenv.mkDerivation {
+    name = "locate-dominating-file";
+    inherit version;
+    src = fetchFromGitHub {
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:15]:
```suggestion
    pname = "locate-dominating-file";
    inherit version;

    src = fetchFromGitHub {
```

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 7]
```diff
@@ -0,0 +1,48 @@
+{
+  bats,
+  bash,
+  lib,
+  stdenv,
+  getopt,
+  fetchFromGitHub,
+}: let
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:8]:
```suggestion
}: 

let
```

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  batsPkg = bats.withLibraries (p: [p.bats-support p.bats-assert]);
+  version = "v0.0.1";
+in
+  stdenv.mkDerivation {
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:12]:
```suggestion
stdenv.mkDerivation {
```

`in` keyword does not start an indentation.

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    name = "locate-dominating-file";
+    inherit version;
+    src = fetchFromGitHub {
+      owner  = "roman";
+      repo   = "locate-dominating-file";
+      rev    = version;
+      sha256 = "sha256-gwh6fAw7BV7VFIkQN02QIhK47uxpYheMk64UeLyp2IY=";
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:19]:
```suggestion
      owner  = "roman";
      repo = "locate-dominating-file";
      rev = "v${version}";
      hash = "sha256-gwh6fAw7BV7VFIkQN02QIhK47uxpYheMk64UeLyp2IY=";
```

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    doCheck = true;
+
+    postPatch = ''
+      for file in $(find src tests -type f); do
+        substituteInPlace "$file" --replace "/bin/bash" "${bash}/bin/bash"
+        substituteInPlace "$file" --replace "#!/usr/bin/env bash" "#!${bash}/bin/bash"
+      done
+    '';
+
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:30]: As I have understood, bats is only used to test. So, add this:

```suggestion

  checkInputs = [
    (bats.withLibraries (p: [ p.bats-support p.bats-assert ]))
  ];

```

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    buildInputs = [getopt];
+
+    installPhase = ''
+      mkdir -p $out/bin
+      cp src/locate-dominating-file.sh $out/bin/locate-dominating-file
+    '';
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:36]:
```suggestion
    installPhase = ''
      runHook preInstall

      mkdir -p $out/bin
      cp src/locate-dominating-file.sh $out/bin/locate-dominating-file
      
      runHook postInstall
    '';
```

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+    checkPhase = ''
+      ${batsPkg}/bin/bats -t tests
+    '';
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:40]:
```suggestion
    checkPhase = ''
      runHook preCheck
      
      bats -t tests
      
      runHook postCheck
    '';
```

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+    meta = with lib; {
+      description = "Program that looks up in a directory hierarchy for a given filename";
+      license = licenses.mit;
+      maintainers = [maintainers.roman];
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:45]:
```suggestion
      maintainers = [ maintainers.roman ];
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/locate-dominating-file/default.nix [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+      rev    = version;
+      sha256 = "sha256-gwh6fAw7BV7VFIkQN02QIhK47uxpYheMk64UeLyp2IY=";
+    };
+
+    doCheck = true;
+
+    postPatch = ''
+      for file in $(find src tests -type f); do
+        substituteInPlace "$file" --replace "/bin/bash" "${bash}/bin/bash"
+        substituteInPlace "$file" --replace "#!/usr/bin/env bash" "#!${bash}/bin/bash"
```
Comment (AndersonTorres) [inline pkgs/development/tools/locate-dominating-file/default.nix:27]:
```suggestion
        patchShebangs "$file"
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by roman** [comment]
Thank you @AndersonTorres, I've learned quite a lot with this review, I pushed new changes (ran `nixpkgs-fmt` instead of `alejandra`) and did the excellent recommendations regarding calling pre/post hooks on the phases and using the `checkInputs`, all in all, great learnings

File: pkgs/tools/misc/locate-dominating-file/default.nix [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  version = "0.0.1";
+in
+stdenvNoCC.mkDerivation {
+  pname = "locate-dominating-file";
+  inherit version;
+  src = fetchFromGitHub {
+    owner  = "roman";
+    repo   = "locate-dominating-file";
+    rev    = version;
+    sha256 = "sha256-gwh6fAw7BV7VFIkQN02QIhK47uxpYheMk64UeLyp2IY=";
```
Comment (AndersonTorres) [inline pkgs/tools/misc/locate-dominating-file/default.nix:18]:
```suggestion
    owner = "roman";
    repo = "locate-dominating-file";
    rev = version;
    sha256 = "sha256-gwh6fAw7BV7VFIkQN02QIhK47uxpYheMk64UeLyp2IY=";
```

No need for pretty tabulation.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Besides, lgtm

File: pkgs/tools/misc/locate-dominating-file/default.nix (hunk 15 above)
Comment (roman) [reply pkgs/tools/misc/locate-dominating-file/default.nix]: :facepalm: I copy pasted that from the previous file without running `nixpkgs-fmt`, done that now

**Review by AndersonTorres — APPROVED** [review]

**Comment by AndersonTorres** [comment]
Now the problem is: why is this failing on all except Arm Linux? Look at the ofborg logs.

**Comment by roman** [comment]
It seems the bats dependency is not building correctly on arm Archs.

What do we do on such situations?

**Comment by AndersonTorres** [comment]
The case is just the reverse: the tests ran OK on Arm, but failed on x86.

My suggestion:

1. Disable tests (`doCheck = false`);
2. Fill a bug report issue about this.

I believe the program can be included into Nixpkgs, nonetheless.

**Comment by roman** [comment]
Hey @AndersonTorres, thank you very much for all the time you have dedicated here; I highly appreciate it.

I did the change for 1.

For (2), I'm not able to find the logs once I force-pushed the commit with the `doCheck = false` :man_facepalming:

Would you know a way to get it back?

File: pkgs/tools/misc/locate-dominating-file/default.nix [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+let
+  version = "0.0.1";
+in
+stdenvNoCC.mkDerivation {
+  pname = "locate-dominating-file";
+  inherit version;
+  src = fetchFromGitHub {
+    owner = "roman";
+    repo = "locate-dominating-file";
+    rev = version;
```
Comment (AndersonTorres) [inline pkgs/tools/misc/locate-dominating-file/default.nix:17]: According to upstream :) it should be
```suggestion
    rev = "v${version}";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/locate-dominating-file/default.nix (hunk 16 above)
Comment (roman) [reply pkgs/tools/misc/locate-dominating-file/default.nix]: :man_facepalming:

File: maintainers/maintainer-list.nix [hunk 17]
```diff
@@ -12241,6 +12241,12 @@
     githubId = 1217934;
     name = "José Romildo Malaquias";
   };
+  roman = {
```
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:12244]: Put this before romildo (alphabetical order).

File: pkgs/tools/misc/locate-dominating-file/default.nix [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+  pname = "locate-dominating-file";
+  inherit version;
+  src = fetchFromGitHub {
+    owner = "roman";
+    repo = "locate-dominating-file";
+    rev = "v${version}";
+    sha256 = "sha256-gwh6fAw7BV7VFIkQN02QIhK47uxpYheMk64UeLyp2IY=";
+  };
+
+  doCheck = false;
```
Comment (AndersonTorres) [inline pkgs/tools/misc/locate-dominating-file/default.nix:21]:
```suggestion
```

File: pkgs/tools/misc/locate-dominating-file/default.nix [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    cp src/locate-dominating-file.sh $out/bin/locate-dominating-file
+
+    runHook postInstall
+  '';
+
```
Comment (AndersonTorres) [inline pkgs/tools/misc/locate-dominating-file/default.nix:39]:
```suggestion

  doCheck = true;
```

Now that the expression is working, let's try ofBorg again and wait its dreadful log...

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by AndersonTorres** [comment]
> Would you know a way to get it back?

Usually ofborg interface shows the builds it tried. You can search them in the tabs.

Or, easier, you can just set `doCheck = true;` again and wait...

File: pkgs/tools/misc/locate-dominating-file/default.nix [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+  pname = "locate-dominating-file";
+  inherit version;
+  src = fetchFromGitHub {
+    owner = "roman";
+    repo = "locate-dominating-file";
+    rev = "v${version}";
+    sha256 = "sha256-gwh6fAw7BV7VFIkQN02QIhK47uxpYheMk64UeLyp2IY=";
+  };
+
+  doCheck = true;
```
Comment (AndersonTorres) [inline pkgs/tools/misc/locate-dominating-file/default.nix:21]: It would be better to put this attribute next to `checkPhase`.

**Review by AndersonTorres — APPROVED** [review]

**Comment by roman** [comment]
Thank for your feedback and time @AndersonTorres, :+1:

**Comment by abathur** [comment]
Hey @roman :)

Nix/nixpkgs doesn't natively understand dependencies in shell scripts, so a straightforward packaging approach like this one leaves the script to pull them out of the user/system environment via PATH. Sometimes that can mean the dependencies aren't available, but it can also mean unexpected versions get pulled in.

While it looks like NixOS systems include GNU `getopt` by default, macOS doesn't and will just pull in the system's default BSD getopt (unless maybe the user has explicitly added the getopt package to their system/user packages). In practice, that looks like:

```console
$ result/bin/locate-dominating-file --help
getopt: illegal option -- n
```

There are a lot of approaches to fixing this, but they basically all boil down to:
- Build a wrapper script that sets or appends the right dependencies before invoking the executable. One easy way to do this is with `makeWrapper` ([doc](https://nixos.org/manual/nixpkgs/stable/#fun-makeWrapper)).
- Directly patch the dependencies in the script so that they refer to store paths. When they already know what the dependencies are or the script is short enough to audit, some people patch them with sed, awk, substituteInPlace, or patsh. 

`makeWrapper` may be fine in your case since this is your own package and you'll probably always know what all needs to get replaced.

I also develop resholve ([doc](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/misc/resholve/README.md)) for the purpose of ferreting out and patching unspecified dependencies in shell scripts by blocking the build until they're all present. Here's an [example of how to do that](https://github.com/abathur/nixpkgs/commit/a452eb84ec159e94a098eed47b01d3cfc2df034b).

<details><summary>And here's a diff to show how it would change the script</summary>

```diff
diff --git a/result/bin/locate-dominating-file b/result-resholved/bin/locate-dominating-file
index 9380ad3b935..6aa3a59468f 100755
--- a/result/bin/locate-dominating-file
+++ b/result-resholved/bin/locate-dominating-file
@@ -3,7 +3,7 @@
 set -euo pipefail
 
 usage() {
-  >&2 cat <<HERE
+  >&2 /nix/store/aldzp4nyc8bh298rdn322vjkbl0rhzvd-coreutils-9.1/bin/cat <<HERE
 Usage:
   locate-dominating-file file|dir [options]
 
@@ -26,7 +26,7 @@ HERE
   exit 2
 }
 
-PARSED_ARGUMENTS="$(getopt -a -n "$(basename "$0")" -o h --long help,print-dir,no-print -- "$@")"
+PARSED_ARGUMENTS="$(/nix/store/qrkz3wlz2yhs1mz1h3hd0kjdr2xwxjwp-getopt-1.1.6/bin/getopt -a -n "$(/nix/store/aldzp4nyc8bh298rdn322vjkbl0rhzvd-coreutils-9.1/bin/basename "$0")" -o h --long help,print-dir,no-print -- "$@")"
 
 VALID_ARGUMENTS=$?
 if [ "$VALID_ARGUMENTS" != "0" ]; then
@@ -63,9 +63,9 @@ locate_dominating_file () {
         if [[ "$NO_PRINT" -eq 1 ]]; then
             true
         elif [[ "$PRINT_FILE" -eq 1 ]]; then
-            echo "$(realpath "$directory")/$FILE_NAME"
+            echo "$(/nix/store/aldzp4nyc8bh298rdn322vjkbl0rhzvd-coreutils-9.1/bin/realpath "$directory")/$FILE_NAME"
         else
-            echo "$(realpath "$directory")"
+            echo "$(/nix/store/aldzp4nyc8bh298rdn322vjkbl0rhzvd-coreutils-9.1/bin/realpath "$directory")"
         fi
         exit 0
     fi
@@ -76,3 +76,10 @@ locate_dominating_file () {
 }
 
 locate_dominating_file
+
+### resholve directives (auto-generated) ## format_version: 3
+# resholve: keep /nix/store/aldzp4nyc8bh298rdn322vjkbl0rhzvd-coreutils-9.1/bin/basename
+# resholve: keep /nix/store/aldzp4nyc8bh298rdn322vjkbl0rhzvd-coreutils-9.1/bin/cat
+# resholve: keep /nix/store/aldzp4nyc8bh298rdn322vjkbl0rhzvd-coreutils-9.1/bin/realpath
+# resholve: keep /nix/store/qrkz3wlz2yhs1mz1h3hd0kjdr2xwxjwp-getopt-1.1.6/bin/getopt
+
```
</details>

With the GNU getopt directly used in the script, it'll work on macOS:

```
$ result-resholved/bin/locate-dominating-file --help
Usage:
  locate-dominating-file file|dir [options]

  Outputs the closest parent directory containing the given file or directory
  name. If it doesn't find file that matches the name, the program exits with
  status code 1.

OPTIONS
  --print-dir  print directory name without file
  --no-print   do not print any message
  --help       prints this message

EXAMPLE

  $ cd /home/roman/my-project/src/internal/lib
  $ locate-dominating-file .env.sh
  => /home/roman/my-project/.env.sh
```

**Comment by roman** [comment]
Hi @abathur,

Thanks for reaching out. 

Coming from a curiosity standpoint, I'm confused about the differences between having `getopt` explicitly as a dependency in the derivation vs using `resholve.mkDerivation`. I noticed on your diff that I still need to list `getopt` in the `solutions` section of the derivation, so why not just include the `getopt` package?

I'll be updating this derivation soon to accommodate the tests in macOS, I'll also try to update the CI in the project to asses if I can have a darwin backend in GHA.

**Comment by abathur** [comment]
> Coming from a curiosity standpoint, I'm confused about the differences between having getopt explicitly as a dependency in the derivation vs using resholve.mkDerivation. I noticed on your diff that I still need to list getopt in the solutions section of the derivation, so why not just include the getopt package?

Not certain exactly how to interpret "having" and "include" here, but maybe I can cover it with a general answer. 

The resholve builder (currently, at least) first runs a normal build with `stdenv.mkDerivation`. Then it runs an outer build using the first as its source, and the `solutions` attr to specify one or more groups of scripts to resolve and how (including what packages to resolve executables from). 

I separate it like this because some packages will have their own build process rather than just being a ready-to-use script; it's easier to let them build normally than try to find the right place to inject resholve in their process without breaking any of their build/check/install/installcheck assumptions. 

resholve could indeed schlorp up whatever is on PATH (including buildInputs), but I elected to make it very explicit to avoid automatically picking up anything unexpected and undocumented. This is a little extra work, but also a little less confusing if, say, someone needs to use resholve on scripts that expect to run variants of common utilities like those in busybox/toybox.

I left getopt in `buildInputs` because the tests failed without it. But now that you've asked the question, I realize I probably should've moved it to checkInputs.

**Pull request merged**